core/src/main/java/com/linecorp/armeria/client/retry/RetryConfig.java

```diff
@@ -0,0 +1,162 @@
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.retry;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.common.Response;
+
+/**
+ * Holds retry config used by a {@link RetryingClient}.
+ * A {@link RetryConfig} instance encapsulates the used {@link RetryRule}, maxTotalAttempts,
+ *  * and responseTimeoutMillisForEachAttempt.
+ */
+public final class RetryConfig<T extends Response> {
+    private final int maxTotalAttempts;
+    private final long responseTimeoutMillisForEachAttempt;
+    @Nullable private final RetryRule retryRule;
+    @Nullable private final RetryRuleWithContent<T> retryRuleWithContent;
+    @Nullable private final RetryRule fromRetryRuleWithContent;
+    private final int maxContentLength;
+    private final boolean needsContentInRule;
+
+    /**
+     * Returns a new {@link RetryConfigBuilder} with the default values from Flags.
+     * Uses a {@link RetryRule}.
+     */
+    public static <T extends Response> RetryConfigBuilder<T> builder(RetryRule retryRule) {
+        return new RetryConfigBuilder<>(retryRule);
+    }
+
+    /**
+     * Returns a new {@link RetryConfigBuilder} with the default values from Flags.
+     * Uses a {@link RetryRuleWithContent} with unlimited content length.
+     */
+    public static <T extends Response> RetryConfigBuilder<T> builder(
+            RetryRuleWithContent<T> retryRuleWithContent) {
+        return new RetryConfigBuilder<>(retryRuleWithContent);
+    }
+
+    /**
+     * Returns a new {@link RetryConfigBuilder} with the default values from Flags.
+     * Uses {@link RetryRuleWithContent} with specified maxContentLength.
+     */
+    public static <T extends Response> RetryConfigBuilder<T> builder(
+            RetryRuleWithContent<T> retryRuleWithContent, int maxContentLength) {
+        return new RetryConfigBuilder<>(retryRuleWithContent, maxContentLength);
+    }
+
+    /**
+     * Returns config's maxTotalAttempt.
+     */
+    public int maxTotalAttempts() {
+        return maxTotalAttempts;
+    }
+
+    /**
+     * Returns config's responseTimeoutMillisForEachAttempt.
+     */
+    public long responseTimeoutMillisForEachAttempt() {
+        return responseTimeoutMillisForEachAttempt;
+    }
+
+    /**
+     * Returns config's retryRule, could be null.
+     */
+    @Nullable
+    public RetryRule retryRule() {
+        return retryRule;
+    }
+
+    /**
+     * Returns config's retryRuleWithContent, could be null.
+     */
+    @Nullable
+    public RetryRuleWithContent<T> retryRuleWithContent() {
+        return retryRuleWithContent;
+    }
+
+    /**
+     * Returns config's retry rule that is converted from retryRuleWithContent, could be null.
+     */
+    @Nullable
+    public RetryRule fromRetryRuleWithContent() {
+        return fromRetryRuleWithContent;
+    }
```

Comment:
How about merging this method into `retryRule()`?

Reply:
The distinction between the regular retryRule and the one converted from retryRuleWithContent is important for calling sites. It is important that retryRule be null when it is not set, and not resemble one converted from the one with content.